File: apps/api/internal/domain/analytics.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// SalesOverview is the response of GET /api/v1/analytics/sales-overview.
type SalesOverview struct {
	TotalBookings  int                `json:"total_bookings"`
	NoShowRate     float64            `json:"no_show_rate"`
	RescheduleRate float64            `json:"reschedule_rate"`
	TotalSales     int                `json:"total_sales"`
	Revenue        string             `json:"revenue"`
	AvgTicket      string             `json:"avg_ticket"`
	ByTag          map[string]int     `json:"by_tag"`
	BySeller       []SalesBySellerRow `json:"by_seller"`
	FunnelDropoff  []FunnelDropoffRow `json:"funnel_dropoff"`
}

type SalesBySellerRow struct {
	SellerID       string  `json:"seller_id"`
	Name           string  `json:"name"`
	Bookings       int     `json:"bookings"`
	Sales          int     `json:"sales"`
	Revenue        string  `json:"revenue"`
	ConversionRate float64 `json:"conversion_rate"`
}

type FunnelDropoffRow struct {
	StepIndex int    `json:"step_index"`
	StepTitle string `json:"step_title"`
	Entered   int    `json:"entered"`
	Exited    int    `json:"exited"`
}

type ABTestForm struct {
	FormID         string  `json:"form_id"`
	Name           string  `json:"name"`
	Starts         int     `json:"starts"`
	Completions    int     `json:"completions"`
	Bookings       int     `json:"bookings"`
	ConversionRate float64 `json:"conversion_rate"`
}

type ABTestResult struct {
	Forms []ABTestForm `json:"forms"`
}

type SalesDailyPoint struct {
	Date      string  `json:"date"`
	Bookings  int     `json:"bookings"`
	Sales     int     `json:"sales"`
	Revenue   string  `json:"revenue"`
}

type AnalyticsEventType string

const (
	EventView                 AnalyticsEventType = "view"
	EventStart                AnalyticsEventType = "start"
	EventQuestionSeen         AnalyticsEventType = "question_seen"
	EventQuestionAnswered     AnalyticsEventType = "question_answered"
	EventBookingSlotSelected  AnalyticsEventType = "booking_slot_selected"
	EventSubmit               AnalyticsEventType = "submit"
	EventAbandon              AnalyticsEventType = "abandon"
	EventShare                AnalyticsEventType = "share"
)

type ResponseEvent struct {
	ID             uuid.UUID          `json:"id"`
	EventID        uuid.UUID          `json:"event_id"`
	FormID         uuid.UUID          `json:"form_id"`
	OrganizationID uuid.UUID          `json:"organization_id"`
	ResponseID     *uuid.UUID         `json:"response_id,omitempty"`
	StepID         *string            `json:"step_id,omitempty"`
	EventType      AnalyticsEventType `json:"event_type"`
	Metadata       json.RawMessage    `json:"metadata"`
	CreatedAt      time.Time          `json:"created_at"`
}

type IngestEventInput struct {
	EventID    uuid.UUID          `json:"event_id"`
	FormID     uuid.UUID          `json:"form_id"`
	ResponseID *uuid.UUID         `json:"response_id,omitempty"`
	StepID     *string            `json:"step_id,omitempty"`
	EventType  AnalyticsEventType `json:"event_type"`
	Metadata   json.RawMessage    `json:"metadata,omitempty"`
}

type IngestBatchInput struct {
	Events []IngestEventInput `json:"events"`
}

type FormDailyMetric struct {
	FormID      uuid.UUID `json:"form_id"`
	Date        string    `json:"date"`
	Views       int       `json:"views"`
	Starts      int       `json:"starts"`
	Completions int       `json:"completions"`
	Abandons    int       `json:"abandons"`
}

type AnalyticsSummary struct {
	Views          int     `json:"views"`
	Starts         int     `json:"starts"`
	Completions    int     `json:"completions"`
	Abandons       int     `json:"abandons"`
	CompletionRate float64 `json:"completion_rate"`
}

type StepDropoff struct {
	StepID    string `json:"step_id"`
	Seen      int    `json:"seen"`
	Answered  int    `json:"answered"`
	DropoffPc float64 `json:"dropoff_pct"`
}

type AnalyticsParams struct {
	FormID uuid.UUID `json:"form_id"`
	From   time.Time `json:"from"`
	To     time.Time `json:"to"`
}

File: apps/web/src/api/endpoints/analytics.ts
import { api } from '@/api/client'

export interface AnalyticsSummary {
  views: number
  starts: number
  completions: number
  abandons: number
  completionRate: number
}

export interface FormDailyMetric {
  formId: string
  date: string
  views: number
  starts: number
  completions: number
  abandons: number
}

export interface StepDropoff {
  stepId: string
  seen: number
  answered: number
  dropoffPct: number
}

export async function getSummary(formId: string, from?: string, to?: string) {
  const params = new URLSearchParams()
  if (from) params.set('from', from)
  if (to) params.set('to', to)
  const qs = params.toString()
  return api<AnalyticsSummary>(`/api/v1/analytics/forms/${formId}/summary${qs ? `?${qs}` : ''}`)
}

export async function getDailyMetrics(formId: string, from?: string, to?: string) {
  const params = new URLSearchParams()
  if (from) params.set('from', from)
  if (to) params.set('to', to)
  const qs = params.toString()
  return api<{ metrics: FormDailyMetric[] }>(`/api/v1/analytics/forms/${formId}/daily${qs ? `?${qs}` : ''}`)
}

export async function getStepDropoff(formId: string, from?: string, to?: string) {
  const params = new URLSearchParams()
  if (from) params.set('from', from)
  if (to) params.set('to', to)
  const qs = params.toString()
  return api<{ steps: StepDropoff[] }>(`/api/v1/analytics/forms/${formId}/dropoff${qs ? `?${qs}` : ''}`)
}

export function getExportUrl(formId: string, from?: string, to?: string) {
  const params = new URLSearchParams()
  if (from) params.set('from', from)
  if (to) params.set('to', to)
  const qs = params.toString()
  return `/api/v1/analytics/forms/${formId}/export${qs ? `?${qs}` : ''}`
}

export async function refreshMetrics() {
  return api<{ status: string }>('/api/v1/analytics/refresh', { method: 'POST' })
}

export async function ingestEvents(events: Array<{
  eventId: string
  formId: string
  responseId?: string
  stepId?: string
  eventType: string
  metadata?: Record<string, unknown>
}>) {
  return api<{ inserted: number }>('/api/v1/public/events', {
    method: 'POST',
    body: { events },
    noAuth: true,
  })
}
